### ISSUE TITLE
Advice on adopting carehare as an alternative to Celery?

### ISSUE THREAD

**Issue by ShaheedHaque** [CJWorkbench/channels_rabbitmq#41]
Hi,

We currently have a decent sized code base that makes use of Celery to distribute CPU intensive and long-lived tasks away from the Django HTTP side of things. When we have code that wants to run under Celery, but from which we wish to drive websockets using channels_rabbitmq, we:

1. Spawn a subprocess.whatever()
2. The subprocess is itself in Python and runs all its code like this:
     
     ```
     def import_all():
          # do everything here
          pass

    loop = asyncio.get_event_loop()
    loop.run_until_complete(sync_to_async(import_all)(*args))
    ```
3. The body of `import_all()` has a loop that broadcasts progress indications to the websocket using a code that boils down in essence to:

    ```
    channel_layer = get_channel_layer()
    async_to_sync(channel_layer.group_send)(group, {'type': 'unicast_progress', 'message': json.dumps(content)})
    ```

I'm aware of the [Publish messages from Celery](https://github.com/CJWorkbench/channels_rabbitmq#publish-messages-from-celery) recipe but given the discussion in #13, as a first step, some time ago I tried unsuccessfully to combine the top level event loop shown at **2** above with this recipe. I just wanted to check if this was a valid approach (that I perhaps just need to persevere with) or whether I'd completely missed the point.

I'm asking the question only now because I have a new subsystem in mind, and was reminded of the suggestion at the bottom of the recipe to avoid Celery and just use carehare directly. I can see how carehare's approach seems more robust as a communication layer with rabbitmq, but I don't think it directly solves the "distributed job queue" problem of distributing work to other cores/hosts? Again, I'm wondering if I just missed something obvious.

**Comment by adamhooper** [comment]
Great question!

If you're trying to send from subprocesses, then you can take two approaches:

1. You can make each subprocess connect to RabbitMQ; or
2. You can make each subprocess pipe its messages back to the parent process ... and the _parent_ process (the Celery process) sends all messages to RabbitMQ.

Personally, I tend to prefer 2. You can `print()` progress messages to stdout in the child process (make sure the environment variable `PYTHONUNBUFFERED=1` is set) and read them in the parent process. The parent-process code to read from a subprocess's stdout isn't trivial; but it makes testing a breeze -- and of course, it avoids unneeded connects+disconnects to RabbitMQ.

You won't be able to access Celery directly, as you're trying. The subprocess isn't a Celery process.

**Comment by ShaheedHaque** [comment]
Stdout/stderr from the subprocess are already used to propagate exceptions across the process boundary, so I'm more inclined to option 1. You didn't rule that out so I assume it's a small matter of reworking that to use carehare {I'm currently pinned to pre-carehare channels rabbitmq). Do you happen to have an example of a simple equivalent to the top level loop I posted?

Also, I think you skipped my second question: so I take it that an actual job queue system based on carehare is an "exercise for the reader"?

**Comment by adamhooper** [comment]
Thanks for replying! Responses inline:

> Stdout/stderr from the subprocess are already used to propagate exceptions across the process boundary, so I'm more inclined to option 1.

I always suggest stderr for errors, stdout for progress reports. I understand reading from two pipes at a time is finicky and verbose in Python ... but it's "correct" in the sense that no sane person would look askance and say, "there must be a simpler way." It's ... well ... _the_ way.

> I assume it's a small matter of reworking that to use carehare

If you're currently pinned to pre-carehare `channels_rabbitmq` and you aren't using `aioamqp`, you should be able to simply unpin and call it a day. Your approach is "supported". (I'll clarify for future readers: it's "supported" in that I'll look into bug reports. But I recommend _not_ running Django Channels in subprocesses, because a stdout approach is easier to implement, maintain and test.)

> I take it that an actual job queue system based on carehare is an "exercise for the reader"?

You can start with one of the examples at https://github.com/CJWorkbench/carehare#usage and grow as needed.

**Comment by ShaheedHaque** [comment]
> I always suggest stderr for errors, stdout for progress reports.

Would that it were so simple. (The dilemma, as you probably know, is that some exceptions which need to be propagated are actually required to be sniffed from logging output AND I want normal output+logs and exception logs to be in one time-ordered sequence plus or minus the vaguaries of buffering)

> If you're currently pinned to pre-carehare `channels_rabbitmq` and you aren't using `aioamqp`, you should be able to simply unpin and call it a day.

I'm not using aioamqp, and that definitely did not work... But I need to try it again to remind myself on why not (and I'll report back on that later if need be).

Thanks again for the considered replies.

**Comment by ShaheedHaque** [comment]
If I take the above code and unpin pin channels_rabbitmq to get 4.0.0, this is what I get:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/channels_rabbitmq/core.py", line 182, in __init__
    asyncio.get_running_loop()
RuntimeError: no running event loop

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/main/srhaque/Innovatieltd/source/paiyroll/views/importer_views.py", line 450, in _import_all
    return xprocess.SubprocessResultsCodec.encode_success(logger, '_import_all', args, body())
  File "/main/srhaque/Innovatieltd/source/paiyroll/views/importer_views.py", line 447, in body
    loop.run_until_complete(sync_to_async(__import_all)(*args))
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/usr/local/lib/python3.8/dist-packages/asgiref/sync.py", line 423, in __call__
    ret = await asyncio.wait_for(future, timeout=None)
  File "/usr/lib/python3.8/asyncio/tasks.py", line 455, in wait_for
    return await fut
  File "/usr/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.8/dist-packages/asgiref/sync.py", line 462, in thread_handler
    return func(*args, **kwargs)
  File "/main/srhaque/Innovatieltd/source/paiyroll/views/importer_views.py", line 477, in __import_all
    progress_notifier(None)
  File "/main/srhaque/Innovatieltd/source/paiyroll/consumers/import_view.py", line 100, in broadcast_progress
    super().broadcast_progress(group, content)
  File "/main/srhaque/Innovatieltd/source/paiyroll/consumers/generic.py", line 83, in broadcast_progress
    channel_layer = get_channel_layer()
  File "/usr/local/lib/python3.8/dist-packages/channels/layers.py", line 359, in get_channel_layer
    return channel_layers[alias]
  File "/usr/local/lib/python3.8/dist-packages/channels/layers.py", line 78, in __getitem__
    self.backends[key] = self.make_backend(key)
  File "/usr/local/lib/python3.8/dist-packages/channels/layers.py", line 44, in make_backend
    return self._make_backend(name, config)
  File "/usr/local/lib/python3.8/dist-packages/channels/layers.py", line 74, in _make_backend
    return backend_class(**config)
  File "/usr/local/lib/python3.8/dist-packages/channels_rabbitmq/core.py", line 184, in __init__
    raise RuntimeError(
RuntimeError: Refusing to initialize channel layer without a running event loop.

If you're writing a django-channels Worker, channels_rabbitmq does not support
this. Read the way to write workers in the carehare documentation.

If you're calling `async_to_sync()`, the call must be within code run by
`sync_to_async()`. Django Channels guarantees this for Django views. Elsewhere,
beware: `asgiref` *does* let you to call `async_to_sync()` without
`sync_to_async()`, but `channels_rabbitmq` *doesn't* -- hence this error.
```
Note that the second traceback starts with an entry from importer_views.py line 450: that's just an artefact of the way the exception from the subprocess is "re-raised" in the parent process (as previously hinted). The next entry correctly shows the traceback as being from this line:
```
    loop.run_until_complete(sync_to_async(__import_all)(*args))
```
The exact sequence of calls in the subprocess is:
```
import_all(some args)                                               <<< start at module scope, i.e. "main".
    return SubprocessResultsCodec.encode_success(..., body())       <<< try/except to check for tracebacks in subprocess.
        loop.run_until_complete(sync_to_async(__import_all)(*args)) <<< body() line 2. Line 1 is "loop = asyncio.get_event_loop()".
```
I should say that none of the functions in this hierarchy are "async def".

What am I doing wrong?

**Comment by adamhooper** [comment]
It's hard to say -- the error isn't exactly in the code you've pasted here, I don't think; and Python asyncio API is a train wreck that leads to days-long bug hunts like this one.

But I can suggest a couple of paths: cleanups I would undertake if I were debugging your code:

1. Use `asyncio.run()`, never `loop.run_until_complete()`. (A RabbitMQ server, like many other servers, will disconnect if the client doesn't respond to heartbeat requests. Pausing the event loop as with `loop.run_until_complete()` may seem like a great idea to event-loop implementors; but for most humans who have jobs to do, it's a disaster.)
2. Define `__import_all()` as async, and rip out all the `async_to_sync()` calls. `sync_to_async()` and `async_to_sync()` use threading: they are far more complex than they appear. (Again: "train wreck" is the appropriate vocabulary.)

TL;DR I expect the bug is in your code, not asgiref/Channels/Django/asyncio. Maybe you're missing an `await` or `async_to_sync()` somewhere? ... but in my experience, the fastest approach to finding the bug is to rewrite the entire codebase to be async. Such is Python.

**Comment by ShaheedHaque** [comment]
Many thanks for the hints...let me go examine my navel...
